flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/input/MultiStateKeyIterator.java

```diff
@@ -46,47 +48,54 @@ public final class MultiStateKeyIterator<K> implements CloseableIterator<K> {
 
     private final KeyedStateBackend<K> backend;
 
-    private final Iterator<K> internal;
+    /** Avoids using Stream#flatMap due to a known flaw, see FLINK-26585 for more details. */
+    private final Iterator<? extends StateDescriptor<?, ?>> outerIter;
+
+    private Iterator<K> innerIter;
 
     private final CloseableRegistry registry;
 
     private K currentKey;
 
     public MultiStateKeyIterator(
             List<? extends StateDescriptor<?, ?>> descriptors, KeyedStateBackend<K> backend) {
+
+        outerIter = descriptors.iterator();
```

Comment:
Move L63-L64 to L68 ?
It's better to call the method after `checkNotNull`

Reply:
Hi @masteryhx , done, rebased, squashed, force pushed as before :)